### PR TITLE
New g4v controller

### PR DESCRIPTION
**Pull request by JaviVilarroig**
New controller for Gemini G4V

### CODE REVIEW

**Comment by uklotzde** [comment]
Test log:
```
[0;32m[ RUN      ] [mControllerPresetValidationTest.MidiPresetsValid
Loading resources from  "/home/travis/build/mixxxdj/mixxx/res/"
"Error parsing XML file /home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml: invalid name for processing instruction at line 3, column 6"
ERROR parsing "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml"
Extension .midi.xml total 113 presets
Extension .hid.xml total 10 presets
Extension .bulk.xml total 1 presets
Validating "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml"
src/test/controller_preset_validation_test.cpp:192: Failure
Value of: preset.isValid()
  Actual: false
Expected: true
LINT: "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml" has no name.
LINT: "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml" has no author.
LINT: "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml" has no description.
LINT: "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml" has no forum link.
LINT: "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml" has no wiki link.
src/test/controller_preset_validation_test.cpp:193: Failure
Value of: lintPresetInfo(preset)
  Actual: false
Expected: true
Searching for controller preset "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml" in paths: "/home/travis/build/mixxxdj/mixxx/res/controllers"
Loading controller preset from "/home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml"
"Error parsing XML file /home/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V.midi.xml: invalid name for processing instruction at line 3, column 6"
src/test/controller_preset_validation_test.cpp:194: Failure
Value of: testLoadPreset(preset)
  Actual: false
Expected: true
```

**Comment by JaviVilarroig** [comment]
Thanks for the feedback @uklotzde :)
I found a minor issue in the XML file. But after updating, I still get errors. AppVeyor seems to fail at apt-get update (!!) and Travis in something related with OSX.
Is there a way to relaunch the tests to discard the possibility of a transient problem in the server?

**Comment by uklotzde** [comment]
Please ignore the AppVeyor timeouts or any infrastructure failures.

But the macOS/clang tests still abort with a crash when trying to load the G4V script in `QScriptEngine::checkSyntax()`? Please note that macOS already uses Qt 5.13 which could make a difference.

```
2019-08-21 11:30:45.002236+0000 mixxx-test[93801:156075] ControllerEngine: Loading "/Users/travis/build/mixxxdj/mixxx/res/controllers/Gemini-G4V-scripts.js"


Process 93801 stopped

* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x187000001858)

    frame #0: 0x00007fff7b7a0724 libsystem_malloc.dylib`tiny_free_list_remove_ptr + 227

libsystem_malloc.dylib`tiny_free_list_remove_ptr:

->  0x7fff7b7a0724 <+227>: movq   %r8, 0x8(%r10)

    0x7fff7b7a0728 <+231>: testq  %r9, %r9

    0x7fff7b7a072b <+234>: jne    0x7fff7b7a073b            ; <+250>

    0x7fff7b7a072d <+236>: jmp    0x7fff7b7a0741            ; <+256>

Target 0: (mixxx-test) stopped.



Process 93801 launched: './mixxx-test' (x86_64)

(lldb) thread backtrace all

* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x187000001858)

  * frame #0: 0x00007fff7b7a0724 libsystem_malloc.dylib`tiny_free_list_remove_ptr + 227

    frame #1: 0x00007fff7b7b5ac8 libsystem_malloc.dylib`tiny_free_no_lock + 1450

    frame #2: 0x00007fff7b7b6254 libsystem_malloc.dylib`free_tiny + 628

    frame #3: 0x00007fff7b7a13bb libsystem_malloc.dylib`szone_realloc + 2545

    frame #4: 0x00007fff7b7a0979 libsystem_malloc.dylib`malloc_zone_realloc + 111

    frame #5: 0x00007fff7b7a086c libsystem_malloc.dylib`realloc + 253

    frame #6: 0x000000010282ea10 QtScript`___lldb_unnamed_symbol3738$$QtScript + 416

    frame #7: 0x00000001027fa3ee QtScript`___lldb_unnamed_symbol2893$$QtScript + 46

    frame #8: 0x00000001027fa3ae QtScript`QScriptEngine::checkSyntax(QString const&) + 14

    frame #9: 0x000000010043fe4c mixxx-test`ControllerEngine::evaluate(this=0x0000000111000800, scriptName=0x0000000106fdeb20, scriptPaths=QList<QString> @ 0x00007ffeefbfde78) at controllerengine.cpp:986

    frame #10: 0x000000010043f285 mixxx-test`ControllerEngine::loadScriptFiles(this=0x0000000111000800, scriptPaths=0x00007ffeefbfe038, scripts=0x00007ffeefbfe098) at controllerengine.cpp:235

    frame #11: 0x0000000100439d8e mixxx-test`Controller::applyPreset(this=0x00007ffeefbfe050, scriptPaths=QList<QString> @ 0x00007ffeefbfe038, initializeScripts=false) at controller.cpp:68

    frame #12: 0x00000001000e0ca1 mixxx-test`ControllerPresetValidationTest::testLoadPreset(this=0x00000001070fa370, preset=0x0000000106fa4040) at controller_preset_validation_test.cpp:143

    frame #13: 0x00000001000e0753 mixxx-test`ControllerPresetValidationTest_MidiPresetsValid_Test::TestBody(this=0x00000001070fa370) at controller_preset_validation_test.cpp:194

    frame #14: 0x000000010086f86e mixxx-test`void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(object=0x00000001070fa370, method=21 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, location="the test body")(), char const*) at gtest.cc:2078

    frame #15: 0x000000010082eb8b mixxx-test`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(object=0x00000001070fa370, method=21 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, location="the test body")(), char const*) at gtest.cc:2114

    frame #16: 0x000000010082ea26 mixxx-test`testing::Test::Run(this=0x00000001070fa370) at gtest.cc:2150

    frame #17: 0x00000001008306ad mixxx-test`testing::TestInfo::Run(this=0x0000000106f11a10) at gtest.cc:2326

    frame #18: 0x0000000100831cbc mixxx-test`testing::TestCase::Run(this=0x0000000106f11b40) at gtest.cc:2444

    frame #19: 0x000000010084278c mixxx-test`testing::internal::UnitTestImpl::RunAllTests(this=0x000000010702ac70) at gtest.cc:4315

    frame #20: 0x00000001008731de mixxx-test`bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=0x000000010702ac70, method=00 24 84 00 01 00 00 00 00 00 00 00 00 00 00 00, location="auxiliary test code (environments or event listeners)")(), char const*) at gtest.cc:2078

    frame #21: 0x00000001008421bb mixxx-test`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=0x000000010702ac70, method=00 24 84 00 01 00 00 00 00 00 00 00 00 00 00 00, location="auxiliary test code (environments or event listeners)")(), char const*) at gtest.cc:2114

    frame #22: 0x000000010084207c mixxx-test`testing::UnitTest::Run(this=0x0000000100da2490) at gtest.cc:3926

    frame #23: 0x0000000100278281 mixxx-test`RUN_ALL_TESTS() at gtest.h:2288

    frame #24: 0x0000000100278240 mixxx-test`main(argc=1, argv=0x00007ffeefbfecc0) at main.cpp:31

    frame #25: 0x00007fff7b5f6115 libdyld.dylib`start + 1

  thread #2

    frame #0: 0x00007fff7b746562 libsystem_kernel.dylib`__workq_kernreturn + 10

    frame #1: 0x00007fff7b88106a libsystem_pthread.dylib`_pthread_wqthread + 1035

    frame #2: 0x00007fff7b880c4d libsystem_pthread.dylib`start_wqthread + 13

  thread #3

    frame #0: 0x00007fff7b746562 libsystem_kernel.dylib`__workq_kernreturn + 10

    frame #1: 0x00007fff7b88106a libsystem_pthread.dylib`_pthread_wqthread + 1035

    frame #2: 0x00007fff7b880c4d libsystem_pthread.dylib`start_wqthread + 13

  thread #4

    frame #0: 0x00007fff7b746562 libsystem_kernel.dylib`__workq_kernreturn + 10

    frame #1: 0x00007fff7b88106a libsystem_pthread.dylib`_pthread_wqthread + 1035

    frame #2: 0x00007fff7b880c4d libsystem_pthread.dylib`start_wqthread + 13

(lldb) script import os; os._exit(1)
```

**Comment by JaviVilarroig** [comment]
I'm having 14 issues in CodeFactor. All of them are about unused variables. This variables come from the function definition but not used in the function. How can I make that issues disappear?

**Comment by uklotzde** [comment]
> I'm having 14 issues in CodeFactor. All of them are about unused variables. This variables come from the function definition but not used in the function. How can I make that issues disappear?

Ignore them. I don't know if marking them with an `_underscore` prefix is a common convention like in Rust.

**Comment by JaviVilarroig** [comment]
Travis is finding a problem in OSX only. It seems related to parsing the JS file.
I have no way to know why it fails.
Can anyone give me a clue on how to diagnose that issue?
Thanks

**Comment by ferranpujolcamins** [comment]
I can reproduce the crash locally:
`malloc: Incorrect checksum for freed object 0x7ff7cf7e4830: probably modified after being freed.`

It happens here ("controllerengine.cpp"):
Screenshot 2019-10-21 at 22 34 29

But the crash itself seems to happen inside QScriptEngine::checkSyntax:
Screenshot 2019-10-21 at 22 36 22

I don't know how to debug further. Any ideas?

**Comment by Be-ing** [comment]
Try this script with #1795. If it works, then I don't think we need to bother looking into this bug.

**Review by Holzhaus — CHANGES_REQUESTED** [review]

**Comment by Holzhaus** [comment]
I propose to move this to the 2.4 milestone. Maybe the crash will be fixed by #1795 but that PR isn't part of milestone 2.3.

**Comment by Be-ing** [comment]
Yes, I agree if #1795 fixes the crash then let's just wait to merge this for 2.4.

**Comment by ferranpujolcamins** [comment]
I tested the scripts on top of  #1795 and the tests pass 👍

**Comment by JaviVilarroig** [comment]
Cool :)
Then we can put this PR in stand-by until #1795 is merged to the master. This will be for 2.4, right?

**Comment by JaviVilarroig** [comment]
PR #1795 has been closed. Can we merge this one?

**Comment by Be-ing** [comment]
#2682 has not been merged yet.

**Comment by JaviVilarroig** [comment]
It seems that #2682 has been merged :)
Can we have this one merged?

**Comment by Be-ing** [comment]
Thank you for the reminder. Can you confirm it works with master now? Also, is the wiki page up to date? https://mixxx.org/wiki/doku.php/gemini_g4v

**Comment by JaviVilarroig** [comment]
The documentation is OK.
I will try tomorrow with master. But the issues where on Mac.

**Comment by JaviVilarroig** [comment]
I can confirm that the controller runs properly on master in my Linux box.
I also fixed a minor change in the FX part. I leave it in the Mappings in Development list. 
I must remember to move it to officially supported when it becomes par of a release.
Is there a common way to have control of the new controllers pending of being released?

**Comment by Be-ing** [comment]
Sorry for letting this slide @JaviVilarroig. I do not have a lot of time now and want to keep helping with the GSOC projects, but I also don't want to keep you waiting as this has already been waiting a long time.

**Comment by JaviVilarroig** [comment]
Don't worry @Be-ing . I agree that the main priority is for the GSOC.
Just wanted to make sure that there was nothing pending on my side.
We are all time stretched.

**Comment by JaviVilarroig** [comment]
Put it on hold.
I recently tested the master branch and found that the controller fails.
I must find some time to check and fix.
:disappointed:

**Comment by Holzhaus** [comment]
Ping.

Btw, starting with Mixxx 2.3, all controller mappings are now supposed to be documented in the Manual. This is necessary so that other users know how to use this mapping. We also need it to review it properly and check if the mapping makes sense and if the code matches the documentation.

Please fork the mixxxdj/manual repository, check out the manual-2.3.x branch and add documentation to the source/hardware/controllers directory. Open the pull request and add the link to the PR description.

Here's an example how such documentation could look like: https://manual.mixxx.org/2.3/en/hardware/controllers/roland_dj_505.html (If you don't have schematic images and don't want to draw them yourself, just use the same numbering as in the controller's manual and document the buttons like that).

**Comment by Holzhaus** [comment]
@JaviVilarroig ping.

**Comment by JaviVilarroig** [comment]
hi @Holzhaus 

Sorry for the delay but my last 4 months have been a disaster. Right now I have no place to live.
I really expect to have solved that in a couple of weeks and being able to start to collaborate again.
Sorry for the inconvenient.

**Comment by JaviVilarroig** [comment]
I have found that my javascript is not working anymore. I works with Mixxx 2.0. Not fully sure, but it seems related with a change in the JS engine in Qt. 
I'm not a big expert in JS, so I'm not sure on how to fix it in an elegant way. I can replace a lot of "this" by object names, but looks too hard coded and not flexible.
When I wrote it (four years ago, more or less) I was not aware of the Components library.
I think that it's better to rewrite the controller using that library than to start tampering with the current code. Do you think this is the right approach?

I'm highly motivated because is my main controller and I really want to be able to use it with the newer versions of Mixxx. That means that I will deliver fast :smile: 

I wonder if it's better to close that PR, so we reduce pollution, and work on my new script and open a new PR when it's ready.

Thanks for your opinion and support!

**Comment by Holzhaus** [comment]
> I'm not a big expert in JS, so I'm not sure on how to fix it in an elegant way. I can replace a lot of "this" by object names, but looks too hard coded and not flexible.

This is what you're looking for: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind

**Comment by Be-ing** [comment]
You don't need to rewrite this from scratch unless you want to. The only breaking changes between 2.3 and main are that `engine.makeConnection`/`engine.connectControl` and `engine.beginTimer` cannot automatically get the `this` object for passed callbacks. They are normal JS callbacks now, so if those callbacks refer to `this`, you need to use `Function.prototype.bind` to bind `this` before passing the function object to Mixxx. So instead of
```
this.conn = engine.makeConnection(this.groupName, this.controlName, this.callbackFunction);
```
do
```
this.conn = engine.makeConnection(this.groupName, this.controlName, this.callbackFunction.bind(this));
```

**Comment by JaviVilarroig** [comment]
Thanks a lot!
It's a pleasure to collaborate with this community.
I will go through that reading and apply the fix.

**Comment by JaviVilarroig** [comment]
Bind did the trick.
I'm going to fully test all the functionality, to make sure there are no other issues.
I have already forked the manual, I will learn how to add the controller documentation :)

**Comment by JaviVilarroig** [comment]
I have managed to fix all the issues that the script was having.
But now, what I have found is that it do not works with my gig setup where I have Mixxx 2.2
Is it OK to commit the changes? I assume yes, because anyway this controller will  only be added to 2.4, if I understood properly.

**Comment by Holzhaus** [comment]
@JaviVilarroig Yes, this PR is targeting the `main` branch, which will eventually become 2.4. If I understood the discussion correctly this mapping won't work on Mixxx <2.4 anyway, right?

Can you run `git pull upstream main` here? Right now the CI build scripts are outdated on this branch.

**Comment by JaviVilarroig** [comment]
Done.
I see that CI has grown a lot!!

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [inline res/controllers/Gemini-G4V-scripts.js:1]: Merge conflict marker?!

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [inline res/controllers/Gemini-G4V-scripts.js:3]: We use eslint now, please remove.

**Review by Holzhaus — CHANGES_REQUESTED** [review]

File: res/controllers/Gemini-G4V-scripts.js
Comment (JaviVilarroig) [reply res/controllers/Gemini-G4V-scripts.js]: Ups. Removed

File: res/controllers/Gemini-G4V-scripts.js
Comment (JaviVilarroig) [reply res/controllers/Gemini-G4V-scripts.js]: done

**Comment by Holzhaus** [comment]
Pre-commit checks fail. Please use our pre-commit hooks to catch issues when committing. That would have also detected the merge conflict markers.

https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking

**Comment by JaviVilarroig** [comment]
I have started to work on the manual.

Is it a prerequisite for having this PR merged?

**Comment by Be-ing** [comment]
Yes. We cannot review controller mappings for hardware we do not have without a description of what the code is intended to do.

**Comment by JaviVilarroig** [comment]
Good copy. I will post a comment when I finish with the manual (another bunch of new technology to learn)

**Comment by JaviVilarroig** [comment]
I already created the PR for the manual page for that controller.

**Comment by Holzhaus** [comment]
Link: https://github.com/mixxxdj/manual/pull/359

**Comment by JaviVilarroig** [comment]
I think that even if the documentation can be improved, we can finalize the review of that PR.
Any comment?

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [inline res/controllers/Gemini-G4V-scripts.js:44]: I think we usually make all global variables/functions/etc. members of a G4V object or something, so avoid conflicts with other js libs. Probably not necessary, but most controller mappings do this:

```javascript
G4V = {}
G4V.Connection = function(...) {
```

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [inline res/controllers/Gemini-G4V-scripts.js:29]: I think we can remove this.

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [inline res/controllers/Gemini-G4V-scripts.js:50]: do we really need this? `print` statements from controller mappings aren't visible anyway unless you pass `--controllerDebug`.

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [inline res/controllers/Gemini-G4V-scripts.js:562]: this looks like you want to use `script.deckFromGroup`.

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [inline res/controllers/Gemini-G4V-scripts.js:281]: this lokks like a constant, please use lowercase param names

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [inline res/controllers/Gemini-G4V-scripts.js:1802]: Please don't do this. That way the error message will only be printed on the terminal (and only if `--controllerDebug` was used) instead of displaying an error popup. The user will be surprised if the controller is not working without warning.

**Review by Holzhaus — COMMENTED** [review]
Looks like you basically reimplemented our midi components library. This is a pretty good indicator that we should work on our documentation and write a proper tutorial for this. This is not the only controller mapping that does this.

I think it's not feasible to rewrite this mapping so I think we may as well merge it, especially since we're going to overhaul our controller system anyway.

@Swiftb0y Do you also want to have a look?

**Comment by Swiftb0y** [comment]
I started a review but didn't get very far. Since this is not supposed to go into 2.3 anyways, I'd like to postpone the merge even more if that is acceptable...

**Comment by JaviVilarroig** [comment]
> Looks like you basically reimplemented our midi components library. This is a pretty good indicator that we should work on our documentation and write a proper tutorial for this. This is not the only controller mapping that does this.
> 
> I think it's not feasible to rewrite this mapping so I think we may as well merge it, especially since we're going to overhaul our controller system anyway.

You're right. I discover the components library after finishing this controller. I wrote my own "library"  when I wrote my first controller for a little Numark controller.

**Comment by JaviVilarroig** [comment]
> I started a review but didn't get very far. Since this is not supposed to go into 2.3 anyways, I'd like to postpone the merge even more if that is acceptable...

Thanks for using your time reviewing my code to both of you. The decision to merge it is on you.

Don't take this as a rant, please. Is not my intention. I started with this PR almost two years ago. I wrote that controller four years ago because I just wanted to use Mixxx with it without any intention to make it official. I also wrote the code for my first controller a Numark Party something like 9 years ago.

I was convinced to open this PR by another Mixxx user that also had a G4V. He was so happy to get it in the forum :)

After this two years I wonder if still has value. Mostly because the G4V start to be old and I'm not sure if it's still in the market. Will be any other user interested on using it?

I know how difficult is to dedicate time to collaborate on a Free Software project and I understand the there are few resources and a lot of things to do. I have limited time to dedicate to that right now due to work situation and personal live (I was granpa recently :) ). I will do my best to fix all the proposed changes as soon as possible but I know there's an schedule. So, if the PR must be postponed do not hesitate.

**Comment by Swiftb0y** [comment]
> After this two years I wonder if still has value. Mostly because the G4V start to be old and I'm not sure if it's still in the market. Will be any other user interested on using it?

I think there is still value in adding it to mixxx. especially for controller which had their support dropped by commercial DJ software.

I understand that its frustrating that this is taking so long. However, the mapping can't go into 2.3 regardless because of the faulty memory access bug occuring in the old QScriptEngine implementation... I'll try to review the code early enough so you have enough time to address my comments when you have the time.

File: res/controllers/Gemini-G4V-scripts.js
Comment (JaviVilarroig) [reply res/controllers/Gemini-G4V-scripts.js:29]: No problem. The prints are more for debugging while  developing. I can remove them.

File: res/controllers/Gemini-G4V-scripts.js
Comment (JaviVilarroig) [reply res/controllers/Gemini-G4V-scripts.js:562]: Most probable. I was not aware of it :)

File: res/controllers/Gemini-G4V-scripts.js
Comment (JaviVilarroig) [reply res/controllers/Gemini-G4V-scripts.js:1802]: Ok. How can I show a meaningful message to the user?

**Comment by JaviVilarroig** [comment]
Just to manage expectations. The proposed changes until now are going to take me some time for coding and testing with the controller.

**Comment by ghomem** [comment]
I am a g4v user and I think supporting it out of the box would add value:-) And by the way: thanks a lot everyone for the efforts.

File: res/controllers/Gemini-G4V-scripts.js
Comment (Holzhaus) [reply res/controllers/Gemini-G4V-scripts.js:1802]: Just don't wrap it in a try catch block. It will still open a popup that loading the controller mapping failed, including a backtrace.

Comment (Holzhaus) [reply res/controllers/Gemini-G4V-scripts.js:1802]: You can test this by throwing an exception in the constructor.

**Comment by JaviVilarroig** [comment]
Hi reviewers.

It seems I'm going to have some spare time to code again after some crazy months.
Looking at all the changes proposed. I'm starting to think that it will be best if I just rewrite it from scratch but using the Mixxx components library.

I just wonder if it's worth. I remember to have read something about a change in the midi controller scripting language (?)

Can you advise?

Thanks

**Comment by Swiftb0y** [comment]
Hey there, I'd greatly appreciate a componentsJS rewrite, it makes the mappings so much easier to review.
In regards to changes in the scripting language, there should not be anything new. In 2.4 we upgraded the environment to be ES7 compatible, but you were already targeting 2.4 so there should not any difference. The only major difference is that we recently enabled the [console API](https://doc.qt.io/qt-5/qtquick-debugging.html#console-api) so make use of the functions there for debugging (for example prefer `console.log` or `console.debug` over our `print`)

**Comment by JaviVilarroig** [comment]
Hi @Swiftb0y 
Ok. I will do that. As this will be a complete rewrite, do you think is better I close this PR and start a new one from zero?

Just to have a clearer history log.

**Comment by Swiftb0y** [comment]
I don't care to be honest. But the only thing you can recycle is probably the XML.

**Comment by JaviVilarroig** [comment]
I'm closing this PR in order to prepare a new one from scratch using the components library.
The associated PR will still be valid. I will use it as a spec set for the new controller.

**Pull request closed without merging**
